File: backend/app/analytics/analytics_service.py
from sqlalchemy.orm import Session

from app.analytics.statistics_service import StatisticsService
from app.analytics.severity_service import SeverityService
from app.analytics.trend_service import TrendService


class AnalyticsService:

    def __init__(self):
        self.statistics = StatisticsService()
        self.severity = SeverityService()
        self.trend = TrendService()

    # ============================================================
    # ANALYTICS SUMMARY
    # ============================================================

    def summary(
        self,
        db: Session,
        user_id
    ) -> dict:

        return {
            "total": self.statistics.total_incidents(
                db,
                user_id=user_id
            ),
            "critical": self.severity.critical(
                db,
                user_id=user_id
            ),
            "high": self.severity.high(
                db,
                user_id=user_id
            ),
            "medium": self.severity.medium(
                db,
                user_id=user_id
            ),
            "low": self.severity.low(
                db,
                user_id=user_id
            )
        }

    # ============================================================
    # RECENT INCIDENTS
    # ============================================================

    def recent_incidents(
        self,
        db: Session,
        user_id
    ) -> dict:

        return {
            "incidents": self.trend.recent(
                db,
                user_id=user_id
            )
        }

    # ============================================================
    # SEVERITY DISTRIBUTION
    # ============================================================

    def severity_distribution(
        self,
        db: Session,
        user_id
    ) -> dict:

        return self.severity.distribution(
            db,
            user_id=user_id
        )

File: frontend/src/services/analyticsService.ts
import api from "../api/api";


// ============================================================
// TYPES
// ============================================================

export interface AnalyticsSummary {
    total: number;
    critical: number;
    high: number;
    medium: number;
    low: number;
}


export interface AnalyticsTrendItem {
    id: number;
    filename: string;
    severity: string;
    created_at: string;
}


export interface AnalyticsTrendResponse {
    incidents: AnalyticsTrendItem[];
}


export interface SeverityDistribution {
    critical: number;
    high: number;
    medium: number;
    low: number;
}


// ============================================================
// SUMMARY
// ============================================================

export async function getAnalyticsSummary(): Promise<
    AnalyticsSummary
> {

    const response =
        await api.get<AnalyticsSummary>(
            "/analytics/summary"
        );

    return response.data;
}


// ============================================================
// RECENT INCIDENTS
// ============================================================

export async function getAnalyticsRecent(): Promise<
    AnalyticsTrendItem[]
> {

    const response =
        await api.get<AnalyticsTrendResponse>(
            "/analytics/recent"
        );

    return response.data.incidents;
}


// ============================================================
// SEVERITY DISTRIBUTION
// ============================================================

export async function getSeverityDistribution(): Promise<
    SeverityDistribution
> {

    const response =
        await api.get<SeverityDistribution>(
            "/analytics/severity"
        );

    return response.data;
}